### PR TITLE
Fix Array.prototype.toString() stackoverflow

### LINKED ISSUE #1975: Calling Array.prototype.toString() can result in a stack overflow exception if the array contains a circular reference.
**Version used**

4.0.3

**Describe the bug**

Calling  `Array.prototype.toString()` or `Array.prototype.join()` produces a Stackoverflow exception when it contains a circular reference.

**To Reproduce**

```cs
 var _engine = new Engine();
 var result = _engine.Evaluate("Array.prototype.toString.call((c = [1, 2, 3, 4], c[1] = c, c))").AsString();
```
**Expected behavior**

```cs
 var _engine = new Engine();
 var result = _engine.Evaluate("Array.prototype.toString.call((c = [1, 2, 3, 4], c[1] = c, c))").AsString();
 Assert.Equal("1,,3,4", result);
```

I expect to get an empty string where the circular reference is set.

**Additional context**
Node and Firefox produce `1,,3,4`

### PR DESCRIPTION
**Pull request by xBaank**
Closes #1975

### CODE REVIEW

File: Jint/Native/Array/ArrayPrototype.cs [hunk 1]
```diff
@@ -1273,14 +1273,14 @@ private JsValue Join(JsValue thisObject, JsValue[] arguments)
             return JsString.Empty;
         }
 
-        static string StringFromJsValue(JsValue value)
+        static string StringFromJsValue(JsValue value, JsValue thisObject)
         {
-            return value.IsNullOrUndefined()
+            return value.IsNullOrUndefined() || thisObject == value
```
Comment (lahma) [inline Jint/Native/Array/ArrayPrototype.cs:1278]: for clarity, I think this should be `ReferenceEquals(thisObject, value)`, the operator overloads can be a bit quirky

**Comment by xBaank** [comment]
I modified the code because `Array.prototype.toString.call((c = [1, 2, 3, 4], b = [1, 2, 3, 4], b[1] = c, c[1] = b, c))` could still produce a StackOverflow. Now it checks if the array is already visited.

**Comment by lahma** [comment]
This will make join calls very costly as now every time another array is being created, or did I misunderstand?

**Comment by xBaank** [comment]
> This will make join calls very costly as now every time another array is being created, or did I misunderstand?

Yes, maybe a better aproach could be to pass down the `visited` hashset along with `thisObject` and `arguments` so no arrays are created?

**Comment by lahma** [comment]
I haven't dug deep into this problem, but I would other types would also be problematic, not on concrete JsArrays if called with different thisObject. Like objects having indexes pointing you circular references, or interop array-like wrappers.

**Comment by lahma** [comment]
By looking at V8 code, it seems that at least following are vulnerable to cyclic behavior (V8 has safeguards):

* `ArrayPrototype.join`
* `ArrayPrototype.toLocaleString`
* `TypedArrayPrototype.join`
* `TypedArrayPrototype.toLocaleString`

**Comment by lahma** [comment]
There's `ObjectTraverseStack` that could probably be used for this, adding `TryEnter` which wouldn't throw like the `Enter` does (`Enter` would `TryEnter` and throw).

**Review by lahma — APPROVED** [review]
Looking good, thank you for taking the time iterating!

**Pull request merged**